### PR TITLE
Update GitHub Actions versions (macOS runners, Windows runners, Node.js 16 actions -> 20)

### PR DESCRIPTION
**Pull request by joshuacwnewton**
## Description

This is a pretty straightforward maintenance PR! 

Relevant GitHub blog posts:

- Node.js 16 deprecated: https://github.blog/changelog/2023-09-22-github-actions-transitioning-from-node-16-to-node-20/
- macos-11 deprecated: https://github.blog/changelog/2024-01-30-github-actions-macos-14-sonoma-is-now-available/
- windows-2022 released: https://github.blog/changelog/2021-11-16-github-actions-windows-server-2022-with-visual-studio-2022-is-now-generally-available-on-github-hosted-runners/

## Linked issues

N/A

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -264,18 +264,18 @@ jobs:
        run: |
          ./.ci.sh -t
 
-  macos-12:
-    name: macOS 12.0 (Monterey)
+  macos-14:
+    name: macOS 14.0 (Sonoma)
```
Comment (joshuacwnewton) [inline .github/workflows/tests.yml:268]: Given that the macOS 14 image is the version that runs on M1 Macs, is it possible that we want to promote this to one of the actions that runs on PRs?

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -396,12 +396,12 @@ jobs:
 
   windows-native:
     name: Windows (Native)
-    runs-on: windows-2019
+    runs-on: windows-2022
```
Comment (joshuacwnewton) [inline .github/workflows/tests.yml:399]: It's possible that we want to selectively keep the old `windows-2019` actions, too, given [this quote](https://github.blog/changelog/2021-11-16-github-actions-windows-server-2022-with-visual-studio-2022-is-now-generally-available-on-github-hosted-runners/):

> The Windows Server 2022 runner image has [different tools and tool versions](https://github.com/actions/virtual-environments/issues/3949) than Windows Server 2019.

**Comment by joshuacwnewton** [comment]
It seems as though all of the tests pass (phew!) -- the current CI failure is due to the upstream issue:

- https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4352

File: .github/workflows/tests.yml (hunk 1 above)
Comment (mguaypaq) [reply .github/workflows/tests.yml:268]: Sure! I expect macOS14 to pick up either installation failures (which don't have to run on every PR) or numerical stability failures (which would be better to have on every PR). For the latter reason, we may also want to have results of `batch_processing.sh` for both macOS 12 and macOS 14 in the next release.

File: .github/workflows/tests.yml (hunk 2 above)
Comment (mguaypaq) [reply .github/workflows/tests.yml:399]: Sure. Unlike macOS 12 vs macOS 14, which has different CPU architecture, I expect Windows 2019 vs 2022 to only be relevant for installation failures. So, it doesn't have to run on every PR, but once in a while wouldn't hurt.

**Review by mguaypaq — APPROVED** [review]
Just one thing: from the commit message of 958cc639410df43c9e92cb6bf7f466aaa4dfbd05:
> setup-wsl@v2 -> not released yet (https://github.com/Vampire/setup-wsl/issues/52)

It looks like `setup-wsl@v3` is now released, so this one can be updated as well.

**Pull request merged**